### ISSUE TITLE
Executing code via a listener during Precognition requests leads to `ERR_CANCELED`

### ISSUE THREAD

**Issue by robdekort** [laravel/precognition#73]
### Laravel Precognition Plugin Version

0.5.5

### Laravel Version

11.10.0

### Plugin

Alpine

### Description

I appreciate this being a complex issue/problem, I'll try to do my best to explain my findings as I do suspect the issue lies somewhere within Precognition. 

I use Statamic together with Precognition (Alpine based) to validate and submit Statamic forms. This works fantastically. No issues there. However as soon as I add a `SubmissionCreated` event to hook into the form submission to custom stuff I run into a problem. I'm creating a payment with a payment provider and this takes a short bit of time. 

If you fill in the form too fast and hit submit before the last validation promise comes back to the client, you get an `ERR_CANCELED`. I suspect this is coming from Axios. You can see this happening here:
https://github.com/laravel/precognition/assets/69107412/0ca281d4-6e06-48a6-aa1a-2e96ee9d39b1

If I properly await the last XHR request before I hit submit you can see everything is fine:
https://github.com/laravel/precognition/assets/69107412/305a0001-faa9-47c0-b985-ccf8e63f0822

So to summarise:
When adding a listener to hook into the form submission to do custom stuff, my `fetch()` submission gets an error back from Precognition before getting the proper validation request back. However, if I remove the form submission listener which takes a bit to execute I cannot reproduce this. 

### Steps To Reproduce

I'd be happy to share access to a private repo with this issue.

**Comment by robdekort** [comment]
I just noticed #69 and especially this comment: https://github.com/laravel/precognition/pull/69#discussion_r1610840065. This might very much be related.

**Comment by timacdonald** [comment]
@robdekort, I think what you are reporting is expected behaviour. Precognition will cancel any in-flight requests to ensure that you don't have "out of order" issues with responses.

However, to make sure I fully understand, could you please share the repo? It would really help me understand the issue better.

**Comment by robdekort** [comment]
Hey @timacdonald, thanks for your reply. I think the same thing. The repo is running Statamic and a private addon that handles the form submission. Is that an issue for you? Happy to talk you through it

**Comment by robdekort** [comment]
Let me see if I can simplify this with just one public repo for you.

**Comment by robdekort** [comment]
Ok, I used a clean Statamic install without fluff, just a precognition form on the homepage. I added instructions for you in the README. Feel free to ping me here or on Discord if you need any assistance. 

https://github.com/robdekort/precognition

**Comment by timacdonald** [comment]
Thanks, @robdekort, gonna dig into this and see if we can get it sorted.

**Comment by robdekort** [comment]
Wonderful. Happy to help if I can assist in any way.

**Comment by timacdonald** [comment]
Also raised in https://github.com/laravel/precognition/issues/74

**Comment by timacdonald** [comment]
@robdekort, I'm not able to replicate the issue in the repo you provided, but I'm wondering if there is a race condition in your code.

Could you try adding the following condition to your `submit` method and see if you notice any difference?

```diff
submit() {
+    if (this.submitted) {
+        return
+    }
+
    this.submitted = true
    this.form.submit()
        .then(response => {
            this.form.reset()
            this.$refs.form.reset()
            this.success = true
            this.submitted = false
            setTimeout(() => {
                this.success = false
            }, 4500)
        })
        .then(this.$refs.form.scrollIntoView())
        .catch(error => {
            const summary = document.querySelector('#summary')
            if (summary) {
                this.$focus.focus(summary.querySelector('a'))
            }
            else {
                console.log(error)
            }
        })
}


If not, can you provide a step by step process to replicate the issue?

**Comment by robdekort** [comment]
Hey @timacdonald. Thanks for your suggestion. Unfortunately that doesn't make a difference. 

To replicate this you need to be quick, so I've turned on throttling in the network tab. And then submit before the last XHR request comes back. To be clear, this is not replicable when there's no listener active.

I've attached a video. I hope this helps. Happy to hop on a call as well to talk you through it.

https://github.com/laravel/precognition/assets/69107412/0b687dcf-e586-48a0-a450-cc736d75c663

**Comment by timacdonald** [comment]
Thanks, @robdekort. This video helped a lot.

I've opened a PR to address this issue: https://github.com/laravel/precognition/pull/78

**Comment by timacdonald** [comment]
I'll close this issue and we can track the PR. If you have any issues, let me know.

**Comment by robdekort** [comment]
Amazing. Looking forward to testing it out. Thank you!
